Add cohorts to BulkEnroll endpoint.

afzaledx opened the pull request:
This PR adds the ability to assign a cohort to the users from the BulkEnroll endpoint.

**JIRA tickets**: 

**Sandbox URL**: https://pr18282.sandbox.opencraft.hosting/

For testing purposes, a course has already been created on the sandbox
https://pr18282.sandbox.opencraft.hosting/courses/course-v1:psX+cs101+2018_T2

Two groups group1 and group2 have been added, and two cohorts cohort1 and cohort2 have been added already.

A token 'fake-token' has been generated for testing with oauth authentication in place. to use this, the token has to be added in the header when using the curl command: 
--header "Authorization: Bearer fake-token"

The example curl calls are all working fine.



**Testing instructions**:
1. Create a new course and add cohorts to it. [cohorts-overview](http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/course_features/cohorts/cohorts_overview.html#cohorts-overview
)

2. hit the bulk enroll endpoint with the correct data and the learner should be enrolled in the course, as well as added to the required cohort.

```
curl -X POST --header "Authorization: Bearer fake-token" -H "Content-Type: application/json" -d '{"action": "enroll","auto_enroll": true,"email_students": true,"courses": "course-v1:psX+cs101+2018_T2","identifiers": "verified", "cohorts": "cohort1"}' https://pr18282.sandbox.opencraft.hosting/api/bulk_enroll/v1/bulk_enroll
```
should return:
```
{"action":"enroll","results":[{"identifier":"verified","after":{"cohort":"cohort1","enrollment":true,"allowed":false,"user":true,"auto_enroll":false},"before":{"cohort":null,"enrollment":false,"allowed":false,"user":true,"auto_enroll":false}}],"auto_enroll":true}},"email_students":true,"auto_enroll":true}
```

An invalid call, where the number of cohorts and number of courses does not match:
```
curl -X POST --header "Authorization: Bearer fake-token" -H "Content-Type: application/json" -d '{"action": "enroll","auto_enroll": true,"email_students": true,"courses": "course-v1:psX+cs101+2018_T2","identifiers": "verified", "cohorts": "cohort1,cohort2"}' https://pr18282.sandbox.opencraft.hosting/api/bulk_enroll/v1/bulk_enroll
```
should return:
```
{"non_field_errors":["If provided, the cohorts and courses should have equal number of items."]}
```



EDXAPP_FEATURES:
  ENABLE_BULK_ENROLLMENT_VIEW: true
  ENABLE_COMBINED_LOGIN_REGISTRATION: true

**Reviewers**
- [ ] (@symbolist)
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @afzaledx! I've created [OSPR-2443](https://openedx.atlassian.net/browse/OSPR-2443) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

afzaledx commented:
jenkins run all

symbolist commented on lms/djangoapps/bulk_enroll/views.py:
```diff
@@ -72,9 +82,23 @@ def post(self, request):
-            for course in serializer.data.get('courses'):
-                response = students_update_enrollment(self.request, course_id=course)
-                response_dict['courses'][course] = json.loads(response.content)
+            for course_id, cohort_name in itertools.izip_longest(serializer.data.get('courses'),
```
This view should raise an error if the number of items in `courses` and `cohorts` are not equal.

afzaledx replied:
Catered for this in the seralizer's validation.

symbolist commented:
@afzaledx This is looking fine. Let me know when you have added the tests and I'll give those a look too.

afzaledx commented:
@symbolist, I have updated the tests. Also catered for the suggestion of unequal cohorts and courses in the serializer's validator.

symbolist commented on lms/djangoapps/bulk_enroll/serializers.py:
```diff
@@ -43,3 +45,19 @@ def validate_courses(self, value):
+        """
+        Check that the cohorts list is the same size as the courses list.
+        """
+        if 'cohorts' in data and len(data['cohorts']) > 0:
```
It is better to be strict with input. I would recommend removing the `len(data['cohorts']) > 0` clause.

afzaledx replied:
Since cohorts is an optional parameter, that will cause all requests that don't contain the cohorts parameter to fail.

symbolist replied:
I meant just making this `if 'cohorts' in data:`. Clients should not be sending in this field if it is going to be empty.

afzaledx replied:
ok. Removed. Will push in a minute.

symbolist commented on lms/djangoapps/bulk_enroll/serializers.py:
```diff
@@ -43,3 +45,19 @@ def validate_courses(self, value):
+        """
+        if 'cohorts' in data and len(data['cohorts']) > 0:
+            if  len(data['cohorts']) != len(data['courses']):
+                raise serializers.ValidationError("If provided, the cohorts and courses should have equal items.")
```
"equal number of items"

symbolist commented on lms/djangoapps/bulk_enroll/views.py:
```diff
@@ -72,9 +83,20 @@ def post(self, request):
+                if cohort_name:
+                    cohort = get_cohort_by_name(course_key=CourseKey.from_string(course_id), name=cohort_name)
+                    for user_data in response_content['results']:
+                        (_, previous_cohort, _) = add_user_to_cohort(cohort, user_data['identifier'])
```
All entries in this list may not be valid enrollments. You need to check that `user_data['after']['enrollment'] is True or user_data['after']['allowed'] is True`.

afzaledx replied:
agreed!

symbolist commented on lms/djangoapps/bulk_enroll/serializers.py:
```diff
@@ -43,3 +45,19 @@ def validate_courses(self, value):
+        """
+        Check that the cohorts list is the same size as the courses list.
+        """
+        if 'cohorts' in data and len(data['cohorts']) > 0:
```
All check that if `cohorts` field is present, `action is "enroll"`.

symbolist commented:
@afzaledx There are a couple of validation checks that need to be added. Otherwise this is mostly looking fine.

symbolist commented:
@afzaledx Also I recommend rebasing your branch from master. There is a `install_node_prereqs` related issue that is causing random test suites to fail.

Let me know when you want me to give a final pass.

afzaledx commented:
@symbolist , I removed the `len(data['cohorts']) > 0` check, and tests started failing. It seems that `'cohorts' in data` is evaluated as True, probably because I added the cohorts field to the serializer.

afzaledx commented:
@symbolist , Yes, I debugged locally, and when running tests which are not passing the cohorts data, the data contains a key with 'cohorts' with an empty list:
OrderedDict([(u'identifiers', u'robot+test+14@edx.org'), (u'courses', [u'org.4/course_4/Run_4']), (u'cohorts', []), (u'action', 'enroll'), (u'auto_enroll', False), (u'email_students', False)])

All the 5 failing tests are having the same issue, so we need both a check for the key as well as the length of the items.

symbolist commented:
@afzaledx Okay. Though you may get a quality failure because of that. You may want to use `if data.get('cohorts'):` in that case (which will evaluate to `False` if `data['cohorts'` is an empty list).

symbolist commented:
@afzaledx As we discussed over the call, there is one more case that will need to be handled and that is if the learner is enrolled in a course and in a cohort, a call is made via this API to set another cohort. I think the learner should be moved over to the new cohort in such a scenario.

afzaledx commented on lms/djangoapps/bulk_enroll/views.py:
```diff
@@ -51,12 +67,19 @@ class BulkEnrollView(APIView):
     """
 
-    authentication_classes = JwtAuthentication, OAuth2Authentication
+    # authentication_classes = JwtAuthentication, OAuth2Authentication
```
Have to remove this. It was removed temporarily to avoid jwt and oauth2 tokens.

afzaledx commented:
@symbolist the sandbox has been updated with the latest code and is ready for testing.

mduboseedx commented:
@afzaledx @symbolist Let me know when this is ready for an edX review. Thanks!

symbolist commented on lms/djangoapps/bulk_enroll/views.py:
```diff
@@ -72,9 +95,32 @@ def post(self, request):
+                    course_key = CourseKey.from_string(course_id)
+                    cohort = get_cohort_by_name(course_key=course_key, name=cohort_name)
+                    for user_data in response_content['results']:
+                        if "after" in user_data and user_data["after"].get("enrollment", False) is True or \
```
This is missing brackets. Should be:
`if "after" in user_data and (user_data["after"].get("enrollment", False) is True or user_data["after"].get("allowed", False) is True):`

symbolist commented:
@afzaledx Besides this one issue, this looks ready!

afzaledx commented:
@symbolist I am having issue with the indentation. I am using pycharm's code formatting but that's failing quality on server. can you tell me what is wrong? views.py line 107/108

symbolist commented:
@afzaledx You can run `pycodestyle <file_path>` to check a specific file.

symbolist commented on lms/djangoapps/bulk_enroll/views.py:
```diff
@@ -72,9 +95,33 @@ def post(self, request):
+                    course_key = CourseKey.from_string(course_id)
+                    cohort = get_cohort_by_name(course_key=course_key, name=cohort_name)
+                    for user_data in response_content['results']:
+                        if "after" in user_data and (
```
```
if "after" in user_data and (
    user_data["after"].get("enrollment", False) is True or
    user_data["after"].get("allowed", False) is True
):
```

symbolist commented on lms/djangoapps/bulk_enroll/views.py:
```diff
@@ -72,9 +95,34 @@ def post(self, request):
+                                add_user_to_cohort(cohort, username)
+                            user_data['after']['cohort'] = cohort_name
+
+                response_dict['courses'][course_id] = response_content
```
(Starting a review)

afzaledx commented:
@mduboseedx, It is ready for edX review now, Thanks.

mduboseedx commented:
@edx/learner Could you take a look at this PR?

pomegranited commented:
Hi @mduboseedx -- could you see when the Learner team might be able to review this PR?  Similar to the others, we'll need a little advanced notice so we can rebase it and re-deploy the sandbox, and set aside time for the upstream review.

staubina commented:
I will check with Product on getting a Product review done.  Once that is complete I will let you know that I am ready to do an engineering review.

staubina commented:
@marcotuts  Can you please review this for Product?

pomegranited commented:
Hi @marcotuts, do you have time to give this a Product review?

pomegranited commented:
Nudge @marcotuts 😄

marcotuts commented:
Would you mind updating this with some screenshot of the specific change? I wasn't able to discern what specifically changed using the sandbox test I did myself. Thanks!

edx-webhook commented:
Hello @afzaledx: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.

edx-webhook commented:
@afzaledx Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18282&entry.1484655318=https://github.com/afzaledx&entry.752974735=2018-05-30+10:06&entry.1917517419=2018-09-24+16:15&entry.2133058324=No) so we can improve your experience in the future.

edx-webhook commented:
@afzaledx Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18282&entry.1484655318=https://github.com/afzaledx&entry.752974735=2018-05-30+10:06&entry.1917517419=2018-09-24+16:15&entry.2133058324=No) so we can improve your experience in the future.

nedbat commented:
@mduboseedx did the bot close this incorrectly?

mduboseedx commented:
@nedbat I suppose it did. Marco accidentally closed this yesterday but then reopened it. I'm not sure why the bot then closed it again.

pomegranited commented:
@nedbat  CC @mduboseedx -- Since @afzaledx has left OpenCraft, we can't edit the PR to make the changes that @marcotuts requested, and closing the PR has also removed our sandbox.

So next week, I'll raise a new, rebased PR with a fresh sandbox, and we can start the upstream review process again.

However, this raises a point -- this doesn't cause any UI changes, so I don't think we need a Product review?  An Educator team review should do.

mduboseedx commented:
@pomegranited I'm actually now meeting with Marco weekly so we will go over any outstanding PRs. AJ from the Learner team had requested that he gives this a look over so he may want that to still happen for your new PR.

Also though, if you have any outstanding PRs definitely let me know. I'll make sure that they are brought up during my meeting with Product.

staubina commented:
Please link the new PR once it is up.

pomegranited commented:
This PR is replaced by https://github.com/edx/edx-platform/pull/19032

CC @mduboseedx  @staubina  @marcotuts
